presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PrefilterForLimitingAggregation.java

```diff
@@ -137,34 +145,34 @@ public PlanNode visitSort(SortNode sortNode, RewriteContext<Void> context)
         @Override
         public PlanNode visitLimit(LimitNode limitNode, RewriteContext<Void> context)
         {
-            PlanNode source = limitNode.getSource();
-            AggregationNode aggregationNode;
+            PlanNode source = rewriteWith(this, limitNode.getSource());
+            AggregationNode aggregationNode = null;
 
             if (source instanceof ProjectNode && ((ProjectNode) source).getSource() instanceof AggregationNode) {
                 aggregationNode = (AggregationNode) ((ProjectNode) source).getSource();
             }
             else if (source instanceof AggregationNode) {
                 aggregationNode = (AggregationNode) source;
             }
-            else {
-                return limitNode;
-            }
 
-            if (!aggregationNode.getGroupingKeys().isEmpty() && isScanFilterProject(aggregationNode.getSource())) {
-                PlanNode rewrittenAggregation = addPrefilter(aggregationNode, limitNode.getCount());
-                if (rewrittenAggregation == aggregationNode) {
-                    return limitNode;
-                }
+            if (aggregationNode != null &&
+                    !aggregationNode.getGroupingKeys().isEmpty()) {
+                Optional<TableScanNode> scanNode = getTableScanNodeWithOnlyFilterAndProject(aggregationNode.getSource());
```

Comment:
can you add a comment here explaining why we only apply for scan/filter/project.